### PR TITLE
Feature/progress bar

### LINKED ISSUE #43: Add flag to tally to restrict cpu intensity

### LINKED ISSUE #44: Add progress bar to `tally` command
Would be fun to see a progress bar for when `tally` is
1. compiling
2. running 

https://github.com/console-rs/indicatif

### PR DESCRIPTION
**Pull request by sivertjoe**
# Changes: 🌟 

This feature (unfortunately) does a couple of things 😅:

1. Refactor how days are being built and executed
2. Add progress bar

## Refactor build and execution steps
The reason I removed futures in favour of threads is that running and executing each day is CPU bound.
I also slowed down on the parallelism. Now at most 25 threads will be opening files.
Doing some quick and dirty benchmarks I found that the old implementation took 10 seconds for `--num-runs 500` while the improved takes 4 seconds. 

## Progress bar
The progress bar looks cool 😎 
Screenshot 2022-12-13 at 18 01 55

Im sure a lot of things are unclear (bad names, unclear functions, etc.) so give me your worst

fixes #44
fixes #43  
^ doesn't exactly do what the issue asks, but I think it solves the main problem: the computer being a hostage to `tally`

### CODE REVIEW

File: src/tally.rs [hunk 1]
```diff
@@ -33,61 +58,73 @@ async fn days() -> Result<(Vec<(PathBuf, u32)>, Vec<u32>), AocError>
...
 
-async fn build_day(day: u32) -> Result<(), AocError>
+async fn build_days(cargo_folder: PathBuf, days: &[(PathBuf, u32)]) -> Result<Vec<u32>, AocError>
 {
-    let bin = format!("day_{:02}", day);
-    let res = tokio::process::Command::new("cargo")
-        .args(["build", "--release", "--bin", &bin])
-        .output()
-        .await?;
+    let sty = ProgressStyle::with_template(
```
Comment (seblj) [inline src/tally.rs:66]: I think we can extract this and the creating of a progress bar to a function that returns the progressbar with the given style. This way we can use that function here and in `run_days`. I think something like:

```rs
fn get_progressbar(len: u64) -> ProgressBar
{
    let sty = ProgressStyle::with_template(
        "[{elapsed_precise}] {msg}... {bar:40.cyan/blue} {pos:>7}/{len:7}",
    )
    .unwrap()
    .progress_chars("##-");

    ProgressBar::new(len).with_style(sty)
}
```

Feel free to change the function name or something, but I think a separate function could be nice though

File: src/tally.rs [hunk 2]
```diff
@@ -33,61 +58,73 @@ async fn days() -> Result<(Vec<(PathBuf, u32)>, Vec<u32>), AocError>
...
+            },
+            Err(_) => Some(day),
+        };
+        progress.inc(1);
+        res
+    });
+    Ok(unimpls.into_iter().flatten().collect())
 }
 
 fn parse_get_times(output: Output) -> Result<(usize, Option<usize>), AocError>
```
Comment (seblj) [inline src/tally.rs:123]: I think we need to be smarter about this (but that can definitely be for another PR IMO). Right now, it considers a day that prints something other than the actual answer as a day that is not implemented. It's not a big problem though as I think when cleaning up and "finishing" up a day, one often removes the other prints 🤷

File: src/tally.rs [hunk 3]
```diff
@@ -33,61 +58,73 @@ async fn days() -> Result<(Vec<(PathBuf, u32)>, Vec<u32>), AocError>
...
+    progress.set_style(sty);
+    progress.set_message("compiling");
 
-async fn build_days(cargo_folder: PathBuf, days: &[(PathBuf, u32)]) -> Result<Vec<u32>, AocError>
-{
-    let mut set = JoinSet::new();
-    for (_path, day) in days.iter()
-    {
+    let res: Result<(), AocError> = thread_exec(days, |(_path, day)| {
         let day = *day;
```
Comment (seblj) [inline src/tally.rs:77]: not needed as `&u32` also implements display

File: src/tally.rs [hunk 4]
```diff
@@ -33,61 +58,73 @@ async fn days() -> Result<(Vec<(PathBuf, u32)>, Vec<u32>), AocError>
...
+    res?;
 
-    let mut vec = Vec::new();
-    for (pb, day) in days
-    {
+    progress.reset();
+    progress.set_message("verifying days");
+
+    let unimpls: Vec<Option<u32>> = thread_exec(days, |(pb, day)| {
         let pb = pb.clone();
```
Comment (seblj) [inline src/tally.rs:100]: clone not needed

File: src/tally.rs [hunk 5]
```diff
@@ -33,61 +58,73 @@ async fn days() -> Result<(Vec<(PathBuf, u32)>, Vec<u32>), AocError>
...
-        }
-    }
+        let bin = format!("day_{:02}", day);
+        let mut target = cargo_folder.clone();
+        target.push("target/release");
+        target.push(&bin);
+        let progress = progress.clone();
 
-    Ok(vec)
+        let res = match std::process::Command::new(target).current_dir(&pb).output()
```
Comment (seblj) [inline src/tally.rs:108]: don't need the reference when we remove the clone as it already is a reference

File: src/tally.rs [hunk 6]
```diff
@@ -33,61 +58,73 @@ async fn days() -> Result<(Vec<(PathBuf, u32)>, Vec<u32>), AocError>
...
 
-    let mut vec = Vec::new();
-    for (pb, day) in days
-    {
+    progress.reset();
+    progress.set_message("verifying days");
+
+    let unimpls: Vec<Option<u32>> = thread_exec(days, |(pb, day)| {
         let pb = pb.clone();
         let day = *day;
```
Comment (seblj) [inline src/tally.rs:101]: maybe just dereference where it is needed in the match(?)

File: src/tally.rs [hunk 7]
```diff
@@ -98,62 +135,42 @@ fn parse_get_times(output: Output) -> Result<(usize, Option<usize>), AocError>
...
     let bin = format!("day_{:02}", day.1);
     let mut target = cargo_folder;
     target.push("target/release");
     target.push(&bin);
+    let mut vec = Vec::with_capacity(number_of_runs);
 
-    let mut set = JoinSet::new();
     for _ in 0..number_of_runs
     {
         let dir = day.0.clone();
```
Comment (seblj) [inline src/tally.rs:153]: can do `let dir = &day.0` instead of clone

File: src/tally.rs [hunk 8]
```diff
@@ -98,62 +135,42 @@ fn parse_get_times(output: Output) -> Result<(usize, Option<usize>), AocError>
...
         let target = target.clone();
```
Comment (seblj) [inline src/tally.rs:154]: can just inline `&target` and get rid of the clone here also

File: src/tally.rs [hunk 9]
```diff
@@ -250,10 +282,7 @@ pub async fn tally(matches: &ArgMatches) -> Result<(), AocError>
     let unimplementeds = build_days(cargo_folder, &have).await?;
 
     have.retain(|elem| !unimplementeds.contains(&elem.1));
-    for unimpl in unimplementeds
-    {
-        dont.push(unimpl);
-    }
+    dont.extend(unimplementeds);
```
Comment (seblj) [inline src/tally.rs:286]: While you are at it 😬

We could get rid of `.cloned()` on `let silver` and `let gold` as well. If we dereference `*p1` inside the map on the line for `let silver = ...`. We can then change the type in `print_info` to expect `Vec<(&u32, usize)>`

File: src/tally.rs [hunk 10]
```diff
@@ -163,21 +180,36 @@ async fn run_days(
...
+        let cargo_folder = cargo_folder.clone();
+        let d = day.1;
 
-    Ok(vec)
+        let res = run_day(cargo_folder, &day, number_of_runs, progress).expect("Running day");
+        (d, res)
+    }))
 }
 
 fn print_info(days: Vec<(u32, (usize, Option<usize>))>, not_done: Vec<u32>, number_of_runs: usize)
```
Comment (seblj) [inline src/tally.rs:215]: Not in the diff here, but I don't think we need to sort the vec created in `print_info`(?)

**Review by seblj — CHANGES_REQUESTED** [review]
Looks really good in general! I think it's more or less ready to go, just some unnecessary clones here and there that could be cleaned up before merging 🚀

File: src/tally.rs (hunk 2 above)
Comment (seblj) [reply src/tally.rs:123]: I'll create an issue for this, and then we can resolve this conversation

File: src/tally.rs (hunk 1 above)
Comment (sivertjoe) [reply src/tally.rs:66]: Good idea

File: src/tally.rs [hunk 11]
```diff
@@ -163,21 +180,36 @@ async fn run_days(
...
+        })
+        .collect::<Vec<_>>();
 
-    while let Some(Ok((day, res))) = set.join_next().await
-    {
-        vec.push((day, res?));
-    }
+    Ok(thread_exec(days, |(day, progress)| {
+        let cargo_folder = cargo_folder.clone();
+        let d = day.1;
```
[the remark this answers is not in the document]
Comment (sivertjoe) [reply src/tally.rs:208]: Inline the thread_exec function?

File: src/tally.rs (hunk 3 above)
Comment (sivertjoe) [reply src/tally.rs:77]: What do you mean?

File: src/tally.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (seblj) [reply src/tally.rs:208]: the `day.1`

File: src/tally.rs (hunk 3 above)
Comment (seblj) [reply src/tally.rs:77]: We don't need the extra `let day = *day` since we only use `day` inside `format!()`. So we don't need the shadowing where we dereference `day`. So we can just use day coming from the closure argument and remove the line `let day = *day`

File: src/tally.rs (hunk 10 above)
Comment (sivertjoe) [reply src/tally.rs:215]: (I renamed _vec to data)
If you mean this line
```rs
let mut data: Vec<_> = vec.iter().map(|(_, time)| *time).collect();
data.sort_unstable();
```
Then I sort it to get the median
```rs
let median = data[data.len() / 2];
```

File: src/tally.rs (hunk 3 above)
Comment (sivertjoe) [reply src/tally.rs:77]: I see, I got confused because I though you referred to 
```rs
let res: Result<(), AocError> = thread_exec(days, |(_path, day)| {
```

File: src/tally.rs (hunk 10 above)
Comment (seblj) [reply src/tally.rs:215]: > Then I sort it to get the median

Ah yes of course, my bad 😅

**Review by seblj — APPROVED** [review]
Ship itttt 🚀

**Comment by sivertjoe** [comment]
:shipit:
